src/meta/src/barrier/mod.rs

```diff
@@ -174,23 +183,64 @@ pub struct GlobalBarrierManager<S: MetaStore> {
 }
 
 struct CheckpointControl<S: MetaStore> {
-    /// Signal whether the barrier with building actor is sent
-    is_build_actor: bool,
     /// Save the state and message of barrier in order
     command_ctx_queue: VecDeque<EpochNode<S>>,
 }
+/// All changed tables of uncommitted barrier
+#[derive(Default)]
+pub struct ChangedTableId {
+    /// In addition to the actors in `Running`.The barrier need to send or collect the actors of
+    /// these tables.
+    create_table_id: Vec<TableId>,
+    /// The barrier does not send or collect the actors of these tables, even if they are `Running`
+    drop_table_id: Vec<TableId>,
+}
+impl ChangedTableId {
+    /// We will return true and send or collect to the actor, if
+    /// 1. `Running` and not be stop by the previous uncommitted barrier
+    /// 2. no `Running` but create by the previous uncommitted barrier or will create by this
+    /// barrier.
+    pub fn can_actor_send_or_collect(&self, s: ActorState, table_id: &TableId) -> bool {
+        s == ActorState::Running && !self.drop_table_id.contains(table_id)
+            || s == ActorState::Inactive && self.create_table_id.contains(table_id)
+    }
+
+    pub fn new<'a, S: MetaStore>(
+        queue_changed_table: impl Iterator<Item = &'a EpochNode<S>>,
+        new_changed_table: ChangedTableState,
+    ) -> Self {
+        let mut create_table_id = vec![];
+        let mut drop_table_id = vec![];
+        queue_changed_table.for_each(|node| match node.command_ctx.command.changed_table_id() {
+            Create(table_id) => create_table_id.push(table_id),
+            Drop(table_id) => drop_table_id.push(table_id),
+            _ => {}
+        });
+        if let Create(table_id) = new_changed_table {
+            create_table_id.push(table_id);
+        };
+        Self {
+            create_table_id,
+            drop_table_id,
+        }
+    }
+}
 
 impl<S> CheckpointControl<S>
 where
     S: MetaStore,
 {
     fn new() -> Self {
         Self {
-            is_build_actor: false,
             command_ctx_queue: VecDeque::default(),
         }
     }
 
+    /// Get the `ChangedTableId` used to modify the actor(send and collect)
+    fn get_changed_table_id(&self, new_changed_table: ChangedTableState) -> ChangedTableId {
+        ChangedTableId::new(self.command_ctx_queue.iter(), new_changed_table)
```

Comment:
Is is possible to have `ChangeTableId` to be part of `CheckpointControl` state so that we can avoid iterating the queue for every new command?

Reply:
Yes, it is. However, we need to maintain two additional queues and some additional work when node are pushed or popped. And we do not have more node. So I iterate the queue for every new command?